### PR TITLE
Re-work stock storage to respect is_in_stock flag.

### PR DESCRIPTION
**Pull request by nikunjkotecha**

### CODE REVIEW

**Comment by nikunjkotecha** [comment]
cc @sdelbosc @vbouchet31

**Comment by malachy-mcconnell** [comment]
This looks like a major feature change. Is it a change or just a refactor to use a separate table and stock entity? Does the LowQuantity flag come from the eCommerce? What about other sock considerations like canBackOrder? Perhaps this is an opportunity to add those additional stock features. Can you summarize any changes to `drush` commands for stick management so that we can update the documentation. Thank you.

**Comment by nikunjkotecha** [comment]
No drush command required now as there is no cache involved and we are currently refactoring to address is_in_stock flag of Magento. Previous drush commands were to handle permanent cache flush / get which is no longer required (https://www.drupal.org/project/pcb / https://github.com/acquia/commerce-manager/pull/118/files#diff-847f41e6f098db399986113e6d5715eeL13)

Till now we were storing only one value in Drupal which was quantity, we update quantity to zero if is_in_stock (or stock status) is 0 (OOS).

Issues we face today on production, we get is_in_stock 0 for parent (configurable) products which we don't consider at all since there is only value available today.

There is a parallel development happening in Magento as well to pass the flag all the time (currently for stock push we do not pass this flag) and also to push for stock status change along with quantity change (so we can get the status of parent as well in Drupal)

Since Magento change is not dependency of changes here code for Drupal is pushed first.

About back-orders, yes after this change we will be able to think about supporting more Magento features, in current project that feature is not used.

Main idea here is to be in-line with upstream system and have stock stored separately (not as part of product data). This allows us to do more then what we do today, complex logic to say if product is in-stock or OOS, invalidate product cache only if product goes from in-stock to OOS or vice versa (today since we update quantity in SKU entity itself caches are in-validated).

**Comment by nikunjkotecha** [comment]
And yes, this is destructive change. (some functions are renamed, some removed, more functions added)

File: modules/acm_sku/acm_sku.info.yml [hunk 1]
```diff
@@ -5,11 +5,11 @@ core: 8.x
 package: Commerce
 dependencies:
   - acquia_commercemanager:acm
+  - acquia_commercemanager:acm_sku_stock
   - key_value_field:key_value_field
   - inline_entity_form:inline_entity_form
   - drupal:menu_ui
   - drupal:node
   - drupal:path
-  - pcb:pcb
```
Comment (miromichalicka) [inline modules/acm_sku/acm_sku.info.yml:13]: It should be removed from composer as well

File: modules/acm_sku/src/StockManager.php [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+use Drupal\Core\Lock\LockBackendInterface;
+use Drupal\Core\Logger\LoggerChannelFactoryInterface;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * Class StockManager.
+ *
+ * @package Drupal\acm_sku
+ */
+class StockManager {
```
Comment (miromichalicka) [inline modules/acm_sku/src/StockManager.php:21]: Why `StockManager` isn't in `acm_sku_stock` submodule? I think it should be contained in that one. Also it should be able to use the rest of suite without `acm_sku_stock` submodule.

File: modules/acm_sku_stock/acm_sku_stock.install [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * @file
+ * Installation and update functions for SKU Stock.
+ */
+
+/**
+ * Implements hook_schema().
+ */
+function acm_sku_stock_schema() {
```
Comment (miromichalicka) [inline modules/acm_sku_stock/acm_sku_stock.install:11]: Why did you decide to use just custom table over custom entity?

File: modules/acm_sku_stock/config/install/acm_sku_stock.settings.yml [hunk 4]
```diff
@@ -0,0 +1 @@
+low_stock: 10
```
Comment (miromichalicka) [inline modules/acm_sku_stock/config/install/acm_sku_stock.settings.yml:1]: This option needs to be exposed in UI.

File: modules/acm_sku/acm_sku.info.yml [hunk 5]
```diff
@@ -5,11 +5,11 @@ core: 8.x
 package: Commerce
 dependencies:
   - acquia_commercemanager:acm
+  - acquia_commercemanager:acm_sku_stock
```
Comment (miromichalicka) [inline modules/acm_sku/acm_sku.info.yml:8]: This submodule shoudn't be dependency for using skus. You're also creating circular dependency, because that submodule depends on this one.

File: modules/acm_sku_stock/acm_sku_stock.info.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+name: Acquia Commerce SKU Stock
+type: module
+description: This module adds stock management functionality to the Acquia Commerce Connector.
+core: 8.x
+package: Commerce
+dependencies:
+  - acm_commerce:acm_sku
```
Comment (miromichalicka) [inline modules/acm_sku_stock/acm_sku_stock.info.yml]: this should be `acquia_commercemanager:acm_sku`

**Review by miromichalicka — CHANGES_REQUESTED** [review]

**Comment by miromichalicka** [comment]
Hi @nikunjkotecha, I briefly went through the code and there are two major problems with it:
1) you shoudn't be forced to use stock
2) there are absolutely no tests. we won't commit this amount of code without proper test coverage.

File: modules/acm_sku/src/StockManager.php (hunk 2 above)
Comment (nikunjkotecha) [reply modules/acm_sku/src/StockManager.php:21]: I was facing issues because of dependencies of it as it is new module introduced now. I looked at changes done in acm modules and followed the approach of ProductManager which is in acm_sku even though new module acm_product was created

File: modules/acm_sku/acm_sku.info.yml (hunk 1 above)
Comment (nikunjkotecha) [reply modules/acm_sku/acm_sku.info.yml:13]: removed

File: modules/acm_sku_stock/acm_sku_stock.install (hunk 3 above)
Comment (nikunjkotecha) [reply modules/acm_sku_stock/acm_sku_stock.install:11]: Entity comes with lot of overhead which we don't want for stock

File: modules/acm_sku/acm_sku.info.yml (hunk 5 above)
Comment (nikunjkotecha) [reply modules/acm_sku/acm_sku.info.yml:8]: aah, true, didn't notice this, removing it from acm_sku_stock module

File: modules/acm_sku_stock/config/install/acm_sku_stock.settings.yml (hunk 4 above)
Comment (nikunjkotecha) [reply modules/acm_sku_stock/config/install/acm_sku_stock.settings.yml:1]: 👍

File: modules/acm_sku_stock/acm_sku_stock.info.yml (hunk 6 above)
Comment (nikunjkotecha) [reply modules/acm_sku_stock/acm_sku_stock.info.yml]: removing this as discussed in another comment

**Comment by nikunjkotecha** [comment]
hi @miromichalicka I'll work on tests as soon as I get chance

about `you shoudn't be forced to use stock`, I think we should address this as separate issue as even current code doesn't support this, thoughts?

File: modules/acm_sku_stock/config/install/acm_sku_stock.settings.yml (hunk 4 above)
Comment (nikunjkotecha) [reply modules/acm_sku_stock/config/install/acm_sku_stock.settings.yml:1]: added

File: modules/acm_sku/acm_sku.info.yml (hunk 5 above)
Comment (nikunjkotecha) [reply modules/acm_sku/acm_sku.info.yml:8]: removed

**Comment by malachy-mcconnell** [comment]
Just a quick note to say the bulk of our tests use Behat so if you can contrive a Behat test for it then that would suffice. (But I don't think our standard template displays stock on the product page; feel free to add stock quantity or low stock warning to the standard templates)

**Comment by nikunjkotecha** [comment]
I was thinking of using phpunit tests, same way as product sync tests, we developed tests for category sync the same way (and I still haven't created PR for it here 😞)

**Comment by nikunjkotecha** [comment]
@miromichalicka added the tests for category sync, thinking of same approach for stock, makes sense to you?

https://github.com/acquia/commerce-manager/pull/119

cc @sdelbosc @vbouchet31

**Comment by miromichalicka** [comment]
@nikunjkotecha yes, I like those category tests, so please go ahead.

**Comment by nikunjkotecha** [comment]
@miromichalicka tests added, ready for final review and merge

File: modules/acm_sku/acm_sku.info.yml (hunk 1 above)
Comment (nikunjkotecha) [reply modules/acm_sku/acm_sku.info.yml:13]: adding back, it is now required by another cache `cache.commerce_user`

**Review by miromichalicka — APPROVED** [review]

**Comment by nikunjkotecha** [comment]
fixed conflicts

**Comment by nikunjkotecha** [comment]
Conflicts resolved and tests passing now.
